File: ui/src/types/product.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

export interface ProductCreate {
  ModelName: string | number;
  Description: string | number;
  CategoryId: number;
  MakeId: number;
  ManufacturingYear: number | null;
  AmcValue: number | null;
}

export interface ProductCreateResult {
  IsProductCreated: Boolean;
}

export const productCreateResultDecoder: Decoder<ProductCreateResult> = object({
  IsProductCreated: boolean,
});

export interface CategoryName {
  Id: number;
  CategoryName: string;
}

export const CategoryNameDecoder: Decoder<CategoryName> = object({
  Id: number,
  CategoryName: string,
});

export interface ProductCategoryNames {
  ProductCategoryNames: CategoryName[];
}

export const ProductCategoryNamesDecoder: Decoder<ProductCategoryNames> = object({
  ProductCategoryNames: array(CategoryNameDecoder),
});

export interface MakeName {
  Id: number;
  Name: string;
}

export const makeNameDecoder: Decoder<MakeName> = object({
  Id: number,
  Name: string,
});

export interface MakeNames {
  MakeNames: MakeName[];
}

export const makesNamesDecoder: Decoder<MakeNames> = object({
  MakeNames: array(makeNameDecoder),
});

export interface ModelName {
  Id: number;
  ModelName: string;
}

export const modelNameDecoder: Decoder<ModelName> = object({
  Id: number,
  ModelName: string,
});

export interface ModelNames {
  ModelNames: ModelName[];
}

export const modelNamesDecoder: Decoder<ModelNames> = object({
  ModelNames: array(modelNameDecoder),
});

export interface ProductDetails {
  Id: number;
  Code: string;
  ModelName: string;
  Description: null | string;
  Category: string | null;
  Make: string | null;
  ManufacturingYear: number;
  AmcValue: number;
  CreatedBy: string;
  CreatedOn: string;
  IsDeleted: Boolean;
}

export interface ProductList {
  Products: ProductDetails[]
  TotalRows: number
  PerPage: number
}

export const productDetailsDecoder: Decoder<ProductDetails> = object({
  Id: number,
  Code: string,
  ModelName: string,
  Description: nullable(string),
  Category: nullable(string),
  Make: nullable(string),
  ManufacturingYear: number,
  AmcValue: number,
  CreatedBy: string,
  CreatedOn: string,
  IsDeleted: boolean,
})

export const productListDecoder: Decoder<ProductList> = object({
  Products: array(productDetailsDecoder),
  TotalRows: number,
  PerPage: number
});

export interface SelectedProductDetails {
  ProductId: number | string;
  ModelName: string;
  Description: string|null;
  AssetProductCategoryId: number | string;
  MakeId: number | string;
  ManufacturingYear: number | string | null;
  AmcValue: number | string | null;
}

export const selectedProductDetailsDecoder: Decoder<SelectedProductDetails> = object({
  ProductId: number,
  ModelName: string,
  Description: nullable(string),
  AssetProductCategoryId: number,
  MakeId: number,
  ManufacturingYear:nullable(number),
  AmcValue: nullable(number)
})

export interface SelectedProduct {
  ProductData: SelectedProductDetails
}

export const selectedProductDecoder: Decoder<SelectedProduct> = object({
  ProductData: selectedProductDetailsDecoder
});

export interface ProductEditResponse {
  isUpdated: Boolean;
}

export const productEditDecoder: Decoder<ProductEditResponse> = object({
  isUpdated: boolean,
});

export interface ProductDeleted {
  IsDeleted: Boolean;
}

export const productDeletedDecoder: Decoder<ProductDeleted> = object({
  IsDeleted: boolean,
});

File: api/BeSureApi/Models/Product.cs
using System.ComponentModel.DataAnnotations;

namespace BeSureApi.Models
{
    public class Product
    {
        public int Id { get; set; }
        public string Code { get; set; }
        public string ModelName { get; set; }
        public string Description { get; set; }
        public string Category { get; set; }
        public string Make { get; set; }
        public short ManufacturingYear { get; set; }
        public int AmcValue { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public bool IsDeleted { get; set; }
    }
    public class ProductCreate
    {
        [Required(ErrorMessage = "validation_error_product_create_model_name_required")]
        [StringLength(64, ErrorMessage = "validation_error_product_create_model_name_max")]
        public string ModelName { get; set; }
        public string Description { get; set; }
        [Required(ErrorMessage = "validation_error_product_create_category_required")]
        public int CategoryId { get; set; }
        [Required(ErrorMessage = "validation_error_product_create_make_required")]
        public int MakeId { get; set; }
       
        public short? ManufacturingYear { get; set; }
        public int? AmcValue { get; set; }
    }
    public class ProductUpdate
    {
        public int ProductId { get; set; }
        [Required(ErrorMessage = "validation_error_product_create_model_name_required")]
        [StringLength(64, ErrorMessage = "validation_error_product_create_model_name_max")]
        public string ModelName { get; set; }
        public string? Description { get; set; }
        [Required(ErrorMessage = "validation_error_product_create_category_required")]
        public int AssetProductCategoryId { get; set; }
        [Required(ErrorMessage = "validation_error_product_create_make_required")]
        public int MakeId { get; set; }
       
        public short ? ManufacturingYear { get; set; }
        public int? AmcValue { get; set; }
        public int UpdatedBy { get; set; }
    }
    public class Deleteproduct
    {
        public int Id { get; set; }
        public int DeletedBy { get; set; }
    }
}
